[Messenger] Improve formatting of exception in failed message

jeroennoten opened the pull request:
| Q             | A
| ------------- | ---
| Branch?       | 5.x
| Bug fix?      | no
| New feature?  | not really, enhancement of an existing feature
| Deprecations? | no
| Tickets       | Fix #32310
| License       | MIT

This PR improves the formatting of exception details in failed messenges when displayed using `messenger:failed:show <id> -vv`.

Before:
<img width="807" alt="Screen Shot 2020-11-01 at 1 05 24 PM" src="https://user-images.githubusercontent.com/4370753/97802602-ea593200-1c44-11eb-8bcb-7fcf2d3f1db0.png">

After:
<img width="803" alt="Screen Shot 2020-11-01 at 1 03 09 PM" src="https://user-images.githubusercontent.com/4370753/97802615-f0e7a980-1c44-11eb-8c12-46b2d4510364.png">

I created a `ThrownExceptionDetails` class which will be displayed as a normal exception when dumped with the VarDumper component. Not sure if this is the right way to do it and if the class is in the right namespace, but this is the best solution I could came up with to fix #32310. I'm open for other suggestions.

jeroennoten committed:
[Messenger] Improve formatting of thrown exception in show failed message command

Nyholm commented on src/Symfony/Component/VarDumper/Caster/ExceptionCaster.php:
```diff
@@ -83,6 +84,15 @@ public static function castThrowingCasterException(ThrowingCasterException $e, a
         return $a;
     }
 
+    public static function castThrownExceptionDetails(ThrownExceptionDetails $exceptionDetails, array $a, Stub $stub, bool $isNested, int $filter = 0)
```
Do we need this method?

Without it it would fallback to `castException`, right? Would that render much differently?

jeroennoten replied:
It will not fallback to `castException`, because `ThrownExceptionDetails` does not extend from `Exception`. Without it, it would not display a nice trace, and the `ThrownExceptionDetails` class will be shown as the name of the exception class, instead of the actual exception class.

Nyholm replied:
Yeah. I just found that out be experimenting. 

Thank you for the explanation.

Nyholm reviewed (commented):
Wow, thank you. I like how simple this solution is. 

Good work

Nyholm commented on src/Symfony/Component/Messenger/Exception/ThrownExceptionDetails.php:
```diff
@@ -0,0 +1,105 @@
+ *
+ * @author Jeroen Noten <jeroennoten@me.com>
+ */
+class ThrownExceptionDetails
```
If this extended `\Exception`, then we would not need to modify the VarDumper component, right?

jeroennoten replied:
We cannot extend `Exception`, because then the file, line and trace details will not be accurate, we cannot set those details to match the actual thrown exception. Note that this class represents an exception that was thrown, stored in the message transport, and then reconstructed to be displayed in the console. This console command has not access to the originally thrown exception.

Nyholm commented on src/Symfony/Component/Messenger/Exception/ThrownExceptionDetails.php:
```diff
@@ -0,0 +1,105 @@
+     */
+    private $trace;
+
+    public function __construct(
```
Let's make this private and accept no arguments.

jeroennoten replied:
Done!

Nyholm commented on src/Symfony/Component/Messenger/Exception/ThrownExceptionDetails.php:
```diff
@@ -0,0 +1,105 @@
+        $this->trace = $trace;
+    }
+
+    public static function createFromFlattenException(FlattenException $flattenException): self
+    {
+        return new self(
+            $flattenException->getClass(),
+            $flattenException->getMessage(),
+            $flattenException->getCode(),
+            $flattenException->getFile(),
+            $flattenException->getLine(),
+            $flattenException->getTrace()
+        );
+    }
```
Let's don't use the constructor with plenty of arguments. 

```suggestion
    public static function createFromFlattenException(FlattenException $flattenException): self
    {
        $model = new self();
        $model->class = $flattenException->getClass();
        $model->message = $flattenException->getMessage();
        $model->code = $flattenException->getCode();
        $model->file = $flattenException->getFile();
        $model->line = $flattenException->getLine();
        $model->trace = $flattenException->getTrace();

        return $model;
    }
```

jeroennoten replied:
Done!

Nyholm reviewed (commented):
Just some minor things.

Nyholm commented:
@yceruto, Is it a good idea to move something like `ThrownExceptionDetails` to the ErrorHandler component?

derrabus reviewed (changes requested):
Please have a look at the Travis php 7.4 run. The failures looks related to your changes.

Nyholm commented:
I think the Travis error in 7.4 is because we need to require symfony/var-dumper:5.3 in messenger's composer.json's require-dev. 

We also need to update branch alias is VarDumps's composer.json.

yceruto commented:
We have a `CliErrorRenderer` class prepared to do this job, maybe we could use it instead of creating another class to format this specific exception. Could you take a look at it?

jeroennoten commented:
> We have a `CliErrorRenderer` class prepared to do this job, maybe we could use it instead of creating another class to format this specific exception. Could you take a look at it?

That's what I tried first. However, the `CliErrorRender` is only capable of rendering `Throwable`s, and we are not able to reconstruct a `Throwable` from the storend data here. I tried adding a method to the `CliErrorRenderer` capable of rendering `FlattenException`, but directly using the VarDumper made more sense to me.

jeroennoten commented:
> I think the Travis error in 7.4 is because we need to require symfony/var-dumper:5.3 in messenger's composer.json's require-dev.
> 
> We also need to update branch alias is VarDumps's composer.json.

@Nyholm that's indeed the reason with the test fails on Travis: the version of symfony/var-dumper used in the tests of symfony/messenger doesn't include the changes I made to the VarDumper component in this PR. I added `"symfony/var-dumper": "^5.3"` to Messenger's require-dev in composer.json and added branch alias `"dev-main": "5.3-dev"` in the VarDumper's composer.json, but it still does not install the version including the changes in this PR. I probably need to replace `dev-main` with something else, but I have no idea with what. Could you help me out?

nicolas-grekas commented on src/Symfony/Component/Messenger/Exception/ThrownExceptionDetails.php:
```diff
@@ -0,0 +1,94 @@
+     */
+    private $line;
+    /**
+     * @var array
```
I'd suggest removing all annotations and replace them with default values.

jeroennoten replied:
Class is removed.

nicolas-grekas commented on src/Symfony/Component/Messenger/Exception/ThrownExceptionDetails.php:
```diff
@@ -0,0 +1,94 @@
+ * file that was distributed with this source code.
+ */
+
+declare(strict_types=1);
```
we don't use strict types

jeroennoten replied:
Class is removed.

nicolas-grekas commented on src/Symfony/Component/VarDumper/composer.json:
```diff
@@ -45,5 +46,10 @@
+    "minimum-stability": "dev",
+    "extra": {
+        "branch-alias": {
+            "dev-main": "5.3-dev"
```
should be removed

jeroennoten replied:
@Nyholm told me to add it. I'm confused now :D How can I make sure that the Travis build uses changes from both packages in the test?

nicolas-grekas replied:
it should be removed, there is no main branch in Symfony
but I get you're both looking for a solution for the CI, which is supposed to pass.
I'm having a look right now.

jeroennoten replied:
It's removed now.

nicolas-grekas commented on src/Symfony/Component/VarDumper/Tests/Caster/ExceptionCasterTest.php:
```diff
@@ -244,4 +244,29 @@ public function testAnonymous()
+
+    public function testThrownExceptionDetails()
+    {
+        $e = \Symfony\Component\ErrorHandler\Exception\FlattenException::createFromThrowable($this->getTestException('foo'));
```
should use `use` statements

jeroennoten replied:
Additions in this class are removed.

nicolas-grekas commented:
> I tried adding a method to the CliErrorRenderer capable of rendering FlattenException, but directly using the VarDumper made more sense to me.

We could use a custom caster attached to the cloner used by the messenger component.
That could prevent adding a new class.
We do this in HttpKernel's DataCollector already.

jeroennoten commented:
> We could use a custom caster attached to the cloner used by the messenger component.

Good suggestion! I'll look into this.

nicolas-grekas commented:
Before pushing the update, can you please just remove the added branch-alias in the component,
and update the root composer.json of symfony/symfony by replacing 5.2 by 5.x in the branch-version entry?
I'd like to see if that will fix the CI.
Thanks.

jeroennoten commented:
> Before pushing the update, can you please just remove the added branch-alias in the component,
> and update the root composer.json of symfony/symfony by replacing 5.2 by 5.x in the branch-version entry?

@nicolas-grekas thanks for you help. Unfortunately, that doesn't seem to work either, see the Travis build.

What do you want me to do? Make other composer.json changes or push the suggested changes that do not involve modifications in the VarDumper componed?

jeroennoten commented:
@nicolas-grekas The changes you pushed seem to work, all Messenger tests pass now. The only thing that is missing to make the VarDumper tests pass as well was a dev-depencency version bump of symfony/messenger to "^5.3" in the VarDumper's composer.json. I added that one.

Let me know when I can remove all unrelated composer.json changes from this MR and update the code to a change in only the Messenger component.

nicolas-grekas commented:
> The changes you pushed seem to work, all Messenger tests pass now

yep, thanks for giving us a good use case to fix this :) This is now merged into 4.4 up to master, you can rebase, and you can also replace the implementation completely with the custom caster.

jeroennoten commented:
Will do! Not sure why the VarDumper tests still cannot find the `ThrownExceptionDetails` class though... Any idea?

nicolas-grekas commented:
> Not sure why the VarDumper tests still cannot find the ThrownExceptionDetails class though... Any idea?

you'd need to add messenger as a dev-dep of var-dumper.

jeroennoten commented:
> > Not sure why the VarDumper tests still cannot find the ThrownExceptionDetails class though... Any idea?
> 
> you'd need to add messenger as a dev-dep of var-dumper.

Right, of course, sorry, was confusing it with the ErrorHandler component for some reason. Thanks!

jeroennoten commented:
@nicolas-grekas I made the changes you suggested resulting in only one changed source file and one changed test file. Could somebody please remove the VarDumper label from this PR?

nicolas-grekas commented on src/Symfony/Component/Messenger/Command/AbstractFailedMessagesCommand.php:
```diff
@@ -172,4 +178,26 @@ protected function getLastRedeliveryStampWithException(Envelope $envelope): ?Red
+
+    private function createCloner(): ?ClonerInterface
+    {
+        if (!class_exists(VarCloner::class) || !class_exists(FlattenException::class)) {
```
`class_exists(FlattenException::class)` is not needed, we don't require the implementation here

jeroennoten replied:
You're right, I removed that check.

nicolas-grekas reviewed (approved):
Much nicer :)
Just one minor suggestion and good to go on my side.

yceruto reviewed (approved):
Thanks!

fabpot commented:
Thank you @jeroennoten.
